.github/scripts/determine-initial-guest-os-versions.py

```diff
@@ -66,36 +70,54 @@ def fetch_versions_from_rollout_dashboard() -> list[str] | None:
     Panics if the parsed data is not in the expected format.
     Returns an empty list if the action is retriable.
     """
-    data = ensure_json_response(ROLLOUT_DASHBOARD_ENDPOINT)
-    if data is None:
-        # Resource was unavailable
-        return []
-
-    versions = set()
     try:
-        for rollout in data:
-            if rollout.get(STATE, None) is None:
-                raise Exception(f"Expected '{STATE}' in 'rollout'")
-            if rollout.get(STATE) in [FAILED, COMPLETE]:
-                continue
-            if rollout.get(BATCHES, None) is None:
-                raise Exception(f"Expected '{BATCHES}' in 'rollout'")
-            for batch in rollout.get(BATCHES).values():
-                if batch.get(SUBNETS, None) is None:
-                    raise Exception(f"Expected '{SUBNETS}' in 'batch'")
-                for subnet in batch.get(SUBNETS):
-                    if subnet.get(GIT_REVISION, None) is None:
-                        raise Exception(f"Expected '{GIT_REVISION}' in 'subnet'")
-                    versions.add(subnet.get(GIT_REVISION))
+        url = ROLLOUT_DASHBOARD_ENDPOINT
+        r =requests.get(url, timeout=30)
+        r.raise_for_status()
+        rollouts = cast(List[Rollout], r.json())
     except Exception as e:
-        # Rollout dashboard returned json but the format changed. Likely an
-        # API change that is not retriable and requires investigation.
-        eprint(f"Error while parsing response from {ROLLOUT_DASHBOARD_ENDPOINT}: {e}")
-        eprint(f"Json response received:\n{data}")
-        exit(1)
-    return list(versions)
+        eprint(f"Error fetching / decoding data from {url}: {e}.  Returning no versions.")
+        return []
 
-def maybe_exectued_timestamp(x : Any) -> int:
+    # The value of the dict entry is datestring, git revision.
+    subnet_to_revision: Dict[str, List[Tuple[str, str]]] = {}
+
+    for rollout in reversed(rollouts):  # Oldest to newest
+        for batch_num_ignored, batch in rollout["batches"].items():
+            for subnet in batch["subnets"]:
+                if subnet["state"] in (SubnetRolloutState.error, SubnetRolloutState.predecessor_failed):
+                    # This subnet failed?  We ignore it, because it could not have been upgraded.
```

Comment:
One refinement we could do here is to test whether a proposal to upgrade this subnet was filed.  This is not currently available in the rollout dashboard API but could be easily made available.

@NikolaMilosa what do you think?

Reply:
I don't think thats needed but feel free to explore if you feel like it